src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -620,9 +622,15 @@ inferDataDeclaration moduleName (ann, tyName, tyArgs, ctors) = do
       let tyCtorName = srcTypeConstructor $ mkQualified tyName moduleName
           tyCtor = foldl (\ty -> srcKindApp ty . srcTypeVar . fst . snd) tyCtorName sigBinders
           tyCtor' = foldl (\ty -> srcTypeApp ty . srcTypeVar . fst) tyCtor tyArgs'
+          vtvs' = S.fromList $ mapMaybe findVtv $ zip tyArgs vtvs where
+            findVtv ((i, _), IsVtaTypeVar) = Just i
+            findVtv _ = Nothing
+          makeVta = everywhereOnTypes go where
+            go (ForAll a b c d e _) | b `S.member` vtvs' = ForAll a b c d e IsVtaTypeVar
+            go t = t
```

Comment:
Similar question here. Additionally, should this function be factored out? It may help if it's possible to go subtlety wrong.